### PR TITLE
95% Working Workspace Loading, Watching, Reference Parsing.

### PR DESCRIPTION
**Pull request by pushqrdx**
This took way more time than I expected, This pr isn't finished as some of the changes need to be discussed.

##### Here's how the new code works:

Project.cs: 
- Handles project loading as well as optional project file watching.
- Provides
    1. Basic info like Name, Path, ProjectFilePath...
    2. NuGetPackageRoot - This one points to the nuget packages folder used by the current project.
    3. List<PackageReference> References:
        - Reference Name
        - Reference Version
        - Reference Path (Based on the NuGetPackageRoot) -- If it's a nuget package.

#### Note: 
The reason i chose to use `NuGetPackageRoot` to find nuget packages is to avoid
having to wait for project builds. Plus some users might have things like `Fody`
or simply chose not to copy the reference dll locally which would result in the server not
being able to get completions for them.

#### Note 2:
See Project.cs#90.

##### And here's how i imagine the server should act:
1. Spin up a process, passing in 1 or more `.csproj` file pathes.
2. Server loads each project file and starts watching it for changes (reference addition, removal, etc).
3. You have the ability to issue subsequent project add, remove in case more projects were added/removed.
4. Each request contains project name as an argument. For example, `getAllCompletions` request might look like:
    ```json
    {
        "Command":"getAllCompletions",
        "Arguments":"AutoUsing"
    }
    ``` 

Feel free to create unit tests for this as i only tested manually, too tired to write tests right now.

### CODE REVIEW

File: AutoUsing/Proxy/Response.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+using Newtonsoft.Json;
+
+namespace AutoUsing
+{
+    /// <summary>
+    /// The JSON serialized response returned to Visual Studio Code. 
+    /// </summary>
+    public abstract class Response<T>
```
Comment (natanfudge) [inline AutoUsing/Proxy/Response.cs]: Explain what is the generic argument (T)

File: AutoUsing/Analysis/DataTypes/PackageReference.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+namespace AutoUsing
+{
+    public class PackageReference
```
Comment (natanfudge) [inline AutoUsing/Analysis/DataTypes/PackageReference.cs]: Document

File: AutoUsing/Analysis/FileWatcher.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+
+using System.IO;
+
+namespace AutoUsing
+{
+    public class FileWatcher : FileSystemWatcher
```
Comment (natanfudge) [inline AutoUsing/Analysis/FileWatcher.cs]: Document

File: AutoUsing/Analysis/Project.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Xml;
+using System.Linq;
+using System;
+
+namespace AutoUsing
+{
+    public class Project : IDisposable
```
Comment (natanfudge) [inline AutoUsing/Analysis/Project.cs]: Document

File: AutoUsing/Analysis/Project.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            GetNuGetRootDirectory();
+
+            // Package References
+            GetPackageReferences();
+
+            // Optional: Watch for changes.
+            if (watch) Watch();
+        }
+
+        private void GetBasicInfo(string filePath)
```
Comment (natanfudge) [inline AutoUsing/Analysis/Project.cs]: I feel that `Get` methods should always return a value. Please rename void `Get` methods to `Load` or something of a kind.

File: AutoUsing/Analysis/Project.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                    Dispose();
+                    return;
+                }
+
+                GetPackageReferences();
+            };
+            FileWatcher.EnableRaisingEvents = true;
+        }
+
+        private void GetNuGetRootDirectory()
```
Comment (natanfudge) [inline AutoUsing/Analysis/Project.cs]: I don't think it's as simple as that, but we'll see

File: AutoUsing/Analysis/Project.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        {
+            Document.Load(Path.Combine(RootDirectory, $"obj/{Name}.csproj.nuget.g.props"));
+
+            NuGetPackageRoot = Document.SelectSingleNode("//x:NuGetPackageRoot", NamespaceManager)?.InnerText;
+        }
+
+        public void GetPackageReferences()
```
Comment (natanfudge) [inline AutoUsing/Analysis/Project.cs]: Change 'get' to 'load' or 'query'

File: AutoUsing/Program.cs [hunk 8]
```diff
@@ -1,64 +1,122 @@
...
+                            }
 
-                        Proxy.WriteData();
-                        break;
-                    case "getAllTypes":
-                        if (Scanner is null) Proxy.WriteData(Errors.LoadAssemblyFirst, false);
+                            // Using C# 7.2 `is expression` to check for null, and assign variable
+                            if (Projects.Find(o => o.Name == projectName) is Project project)
+                            {
+                                // ? Wanna determine whether we scan assemblies on the fly
```
Comment (natanfudge) [inline AutoUsing/Program.cs]: See PR comment.

File: AutoUsing/Proxy/Response.cs (hunk 1 above)
Comment (pushqrdx) [reply AutoUsing/Proxy/Response.cs]: For the implicit operator to be able to serialize a JSON objects for classes that inherit Response.

Comment (natanfudge) [reply AutoUsing/Proxy/Response.cs]: (in the xml doc)

**Pull request closed without merging**